prometheus/go_collector.go

```diff
@@ -16,16 +19,31 @@ func NewGoCollector() *goCollector {
 			Name: "process_goroutines",
 			Help: "Number of goroutines that currently exist.",
 		}),
+		gcDesc: NewDesc(
+			"process_gc_duration_seconds",
+			"A summary of the GC invocation durations.",
+			nil, nil),
 	}
 }
 
 // Describe returns all descriptions of the collector.
 func (c *goCollector) Describe(ch chan<- *Desc) {
 	ch <- c.goroutines.Desc()
+	ch <- c.gcDesc
 }
 
 // Collect returns the current state of all metrics of the collector.
 func (c *goCollector) Collect(ch chan<- Metric) {
 	c.goroutines.Set(float64(runtime.NumGoroutine()))
 	ch <- c.goroutines
+
+	var stats debug.GCStats
+	stats.PauseQuantiles = make([]time.Duration, 4)
+	debug.ReadGCStats(&stats)
+
+	quantiles := make(map[float64]float64)
+	for idx, pq := range stats.PauseQuantiles {
+		quantiles[float64(idx+1)/float64(len(stats.PauseQuantiles))] = pq.Seconds()
```

Comment:
As per https://golang.org/pkg/runtime/debug/#ReadGCStats, the first entry in `stats.PauseQuantiles` is the minimum, while the last one is the maximum (so, 0th and 100th percentiles, losely speaking). The full contained quantiles in `stats.PauseQuantiles` when choosing a length of 4 will be: minimum, 0.333..., 0.666..., maximum.

The code here sets the first entry to the 25th percentile ((0+1) / 4 == 0.25), and so on, not quite matching the actual returned quantiles. It should probably follow the code at https://golang.org/src/runtime/debug/garbage.go?s=2860:3040 to get the right quantiles.

Also, rather than 0.333..., 0.666..., how about making `stats.PauseQuantiles` with a length of 5, which would give us: minimum, 0.25, 0.5, 0.75, maximum. Not sure if we should actually end up recording the minimum and maximum as the `0` and `1` quantiles. @beorn7, @brian-brazil?